packages/run-protocol/src/collect.js

```diff
@@ -1,29 +1,67 @@
 // @ts-check
 
-const { entries, fromEntries, keys, values } = Object;
-
-export const Collect = {
-  /**
-   * @param {Record<string, V>} obj
-   * @param {(v: V) => U} f
-   * @returns {Record<string, U>}
-   * @template V
-   * @template U
-   */
-  mapValues: (obj, f) => fromEntries(entries(obj).map(([p, v]) => [p, f(v)])),
-  /**
-   * @param {X[]} xs
-   * @param {Y[]} ys
-   * @returns {[X, Y][]}
-   * @template X
-   * @template Y
-   */
-  zip: (xs, ys) => xs.map((x, i) => [x, ys[i]]),
-  /**
-   * @param {Record<string, ERef<V>>} obj
-   * @returns {Promise<Record<string, V>>}
-   * @template V
-   */
-  allValues: async obj =>
-    fromEntries(Collect.zip(keys(obj), await Promise.all(values(obj)))),
-};
+import { deeplyFulfilled } from '@endo/marshal';
+
+// const { entries, fromEntries, keys, values } = Object;
+const { entries, fromEntries } = Object;
+
+// Why the `Collect` object? In any case, it was missing a `harden` or `Far`.
+// But why not just export the useful functions directly?
+
+// Huh. I didn't know you could put the @template at the bottom.
+// See `objectMap` in objArrayConversion.js
+// Read the qualifiers in the comment!
+/**
+ * @param {Record<string, V>} obj
+ * @param {(v: V) => U} f
+ * @returns {Record<string, U>}
+ * @template V
+ * @template U
+ */
+export const mapValues = (obj, f) =>
+  harden(fromEntries(entries(obj).map(([p, v]) => [p, f(v)])));
+harden(mapValues);
+
+/**
+ * @param {X[]} xs
+ * @param {Y[]} ys
+ * @returns {[X, Y][]}
+ * @template X
+ * @template Y
+ */
+export const zip = (xs, ys) => harden(xs.map((x, i) => [x, ys[i]]));
+harden(zip);
+
+// Are you sure you only want one level?
+// See deeplyFulfilled in deeplyFulfilled.js
+// /**
+//  * @param {Record<string, ERef<V>>} obj
+//  * @returns {Promise<Record<string, V>>}
+//  * @template V
+//  */
+// export const allValues = async obj => {
+//   // Get both keys and values before the `await`, so they're consistent
+//   // (assuming that `obj` is not a proxy, which we don't validate)
+//   const ks = keys(obj);
+//   const vs = await Promise.all(values(obj));
+//   return harden(fromEntries(zip(ks, vs)));
+// };
+// harden(allValues);
+
+// This should have worked, but was passed something non-hardened
+// export const allValues = deeplyFulfilled;
+// So I tried the following instead
+export const allValues = obj => deeplyFulfilled(harden(obj));
```

Comment:
I left this uncommented for now, and lines 37-49 commented above, so the resulting failures will appear under CI. See comment below for the bug causing these failures.

Reply:
Note that this `harden` masks, rather than fixes, bugs elsewhere. The passing around of those unhardened objects may have already caused problems before they reach here.